controller/localdevice/finalizer.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2020 The MayaData Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package localdevice
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"openebs.io/metac/controller/generic"
+
+	ccc "mayadata.io/cstorpoolauto/util/cstorclusterconfig"
+	metacutil "mayadata.io/cstorpoolauto/util/metac"
+)
+
+type finalizer struct {
+	request  *generic.SyncHookRequest
+	response *generic.SyncHookResponse
+
+	isDiskLocal bool
+	fatal       error
+	err         error
+	warns       []string
+}
+
+func (s *finalizer) validateArgs() {
+	// validation failure of request &/ response is a fatal error
+	s.fatal = metacutil.ValidateGenericControllerArgs(s.request, s.response)
+}
+
+func (s *finalizer) skipIfNotLocalDisk() {
+	s.isDiskLocal, s.err =
+		ccc.NewHelper(s.request.Watch).IsLocalBlockDiskConfig()
+	if s.err != nil {
+		return
+	}
+	if !s.isDiskLocal {
+		var msg = fmt.Sprintf(
+			"Will skip LocalDevice finalize: DiskConfig is not local: Watch %q - %q / %q",
+			s.request.Watch.GetKind(),
+			s.request.Watch.GetNamespace(),
+			s.request.Watch.GetName(),
+		)
+		glog.V(3).Infof(msg)
+		// we want to skip reconciling this since this CstorClusterConfig
+		// is not meant for local devices
+		s.response.SkipReconcile = true
+	}
+}
+
+func (s *finalizer) logFinalizeStart() {
```

Comment:
Why we want to have a log as a function

Reply:
Since we are executing each logic as a function, log needs to be treated as a function. This may not be idiomatic & might look bad, however this is the best I could get.